#Exercise 2

In [1]:
!pip install -q pyomo

In [2]:
from pyomo.environ import *

In [3]:
import numpy as np

#Ans 1:

Decision Variables : $x_{ij}=\left\{
\begin{array}{ll}
      1 & \text{when facility i is located at location j} \\
      0 & \text{when facility i is not located at location j} \\
\end{array}\right.\\i = 1,\cdots,12\\j = 1,\cdots,12$

$\text{min } \sum_{i=1}^{12}\sum_{j=1}^{12}c_{ij}x_{ij}$

subject to:

$\bullet\sum_{i=1}^{12}x_{ij}=1 \quad \forall \quad j = 1,\cdots,12\\
\bullet\sum_{j=1}^{12}x_{ij}=1 \quad \forall \quad i = 1,\cdots,12\\
\bullet x_{ij}=\{0,1\}$

In [4]:
coef = np.loadtxt('/content/lab5_ex2.txt', delimiter = ',')

In [5]:
print(coef.shape)

(12, 12)


In [6]:
model = ConcreteModel()

In [7]:
M = coef.shape[0]
N = coef.shape[1]

In [8]:
row_indices = np.arange(M)
col_indices = np.arange(N)

In [9]:
model.x = Var(row_indices,col_indices,domain=Binary)

In [10]:
model.objective = Objective(expr=sum(coef[i][j]*model.x[i,j] for i in row_indices for j in col_indices),sense = minimize)

In [11]:
model.constraints = ConstraintList()

In [12]:
for i in row_indices:
  model.constraints.add(sum(model.x[i,j] for j in col_indices) == 1)
for j in col_indices:
  model.constraints.add(sum(model.x[i,j] for i in row_indices) == 1)

In [13]:
model.pprint()

4 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   24 : {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain              : Size : Members
        None :     2 : x_index_0*x_index_1 :  144 : {(0, 0), (0, 1), (0, 2), (0, 3), (0, 4), (0, 5), (0, 6), (0, 7), (0, 8), (0, 9), (0, 10), (0, 11), (1, 0), (1, 1), (1, 2), (1, 3), (1, 4), (1, 5), (1, 6), (1, 7), (1, 8), (1, 9), (1, 10), (1, 11), (2, 0), (2, 1), (2, 2), (2, 3), (2, 4), (2, 5), (2, 6), (2, 7), (2, 8), (2, 9), (2, 10), (2, 11), (3, 0), (3, 1), (3, 2), (3, 3), (3, 4), (3, 5), (3, 6), (3, 7), (3, 8), (3, 9), (3, 10), (3, 11), (4, 0), (4, 1), (4, 2), (4, 3), (4, 4), (4, 5), (4, 6), (4, 7), (4, 8), (4, 9), (4, 10), (4, 11), (5, 0), (5, 1), (5, 2), (5, 3), (5, 4), (5, 5), (5, 6), (5, 7), (5, 8), (5, 9), (5, 10), (5, 1

In [14]:
!apt-get install -y -qq coinor-cbc

In [15]:
opt_cbc = SolverFactory('cbc')

In [16]:
result = opt_cbc.solve(model)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [17]:
print('Objective = ', model.objective())

print('\nDecision Variables')
for i in row_indices:
  for j in col_indices:
    if model.x[i,j].value == 1:
      print('x[',i,'][',j,']:',model.x[i,j].value)

print('\nConstraints')
model.constraints.display()

Objective =  203.0

Decision Variables
x[ 0 ][ 6 ]: 1.0
x[ 1 ][ 1 ]: 1.0
x[ 2 ][ 4 ]: 1.0
x[ 3 ][ 0 ]: 1.0
x[ 4 ][ 10 ]: 1.0
x[ 5 ][ 2 ]: 1.0
x[ 6 ][ 5 ]: 1.0
x[ 7 ][ 3 ]: 1.0
x[ 8 ][ 9 ]: 1.0
x[ 9 ][ 11 ]: 1.0
x[ 10 ][ 8 ]: 1.0
x[ 11 ][ 7 ]: 1.0

Constraints
constraints : Size=24
    Key : Lower : Body : Upper
      1 :   1.0 :  1.0 :   1.0
      2 :   1.0 :  1.0 :   1.0
      3 :   1.0 :  1.0 :   1.0
      4 :   1.0 :  1.0 :   1.0
      5 :   1.0 :  1.0 :   1.0
      6 :   1.0 :  1.0 :   1.0
      7 :   1.0 :  1.0 :   1.0
      8 :   1.0 :  1.0 :   1.0
      9 :   1.0 :  1.0 :   1.0
     10 :   1.0 :  1.0 :   1.0
     11 :   1.0 :  1.0 :   1.0
     12 :   1.0 :  1.0 :   1.0
     13 :   1.0 :  1.0 :   1.0
     14 :   1.0 :  1.0 :   1.0
     15 :   1.0 :  1.0 :   1.0
     16 :   1.0 :  1.0 :   1.0
     17 :   1.0 :  1.0 :   1.0
     18 :   1.0 :  1.0 :   1.0
     19 :   1.0 :  1.0 :   1.0
     20 :   1.0 :  1.0 :   1.0
     21 :   1.0 :  1.0 :   1.0
     22 :   1.0 :  1.0 :   1.0
     

#Ans 8:

Facility 1 must be opened at location 7.

Facility 2 must be opened at location 2.

Facility 3 must be opened at location 5.

Facility 4 must be opened at location 1.

Facility 5 must be opened at location 11.

Facility 6 must be opened at location 3.

Facility 7 must be opened at location 6.

Facility 8 must be opened at location 4.

Facility 9 must be opened at location 10.

Facility 10 must be opened at location 12.

Facility 11 must be opened at location 9.

Facility 12 must be opened at location 8.

In [18]:
model_reals = ConcreteModel()

In [19]:
model_reals.y = Var(row_indices,col_indices,domain=NonNegativeReals)

In [20]:
model_reals.objective = Objective(expr=sum(coef[i][j]*model_reals.y[i,j] for i in row_indices for j in col_indices),sense = minimize)

In [21]:
model_reals.constraints = ConstraintList()
for i in row_indices:
  model_reals.constraints.add(sum(model_reals.y[i,j] for j in col_indices) == 1)
for j in col_indices:
  model_reals.constraints.add(sum(model_reals.y[i,j] for i in row_indices) == 1)

In [22]:
model_reals.pprint()

4 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   24 : {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24}
    y_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain              : Size : Members
        None :     2 : y_index_0*y_index_1 :  144 : {(0, 0), (0, 1), (0, 2), (0, 3), (0, 4), (0, 5), (0, 6), (0, 7), (0, 8), (0, 9), (0, 10), (0, 11), (1, 0), (1, 1), (1, 2), (1, 3), (1, 4), (1, 5), (1, 6), (1, 7), (1, 8), (1, 9), (1, 10), (1, 11), (2, 0), (2, 1), (2, 2), (2, 3), (2, 4), (2, 5), (2, 6), (2, 7), (2, 8), (2, 9), (2, 10), (2, 11), (3, 0), (3, 1), (3, 2), (3, 3), (3, 4), (3, 5), (3, 6), (3, 7), (3, 8), (3, 9), (3, 10), (3, 11), (4, 0), (4, 1), (4, 2), (4, 3), (4, 4), (4, 5), (4, 6), (4, 7), (4, 8), (4, 9), (4, 10), (4, 11), (5, 0), (5, 1), (5, 2), (5, 3), (5, 4), (5, 5), (5, 6), (5, 7), (5, 8), (5, 9), (5, 10), (5, 1

In [23]:
result_reals = opt_cbc.solve(model_reals)
print('Solver status:', result_reals.solver.status)
print('Solver termination condition:',result_reals.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [24]:
print('Objective = ', model_reals.objective())

print('\nDecision Variables')
for i in row_indices:
  for j in col_indices:
    if model.x[i,j].value == 1:
      print('x[',i,'][',j,']:',model_reals.y[i,j].value)

print('\nConstraints')
model_reals.constraints.display()

Objective =  203.0

Decision Variables
x[ 0 ][ 6 ]: 1.0
x[ 1 ][ 1 ]: 1.0
x[ 2 ][ 4 ]: 1.0
x[ 3 ][ 0 ]: 1.0
x[ 4 ][ 10 ]: 1.0
x[ 5 ][ 2 ]: 1.0
x[ 6 ][ 5 ]: 1.0
x[ 7 ][ 3 ]: 1.0
x[ 8 ][ 9 ]: 1.0
x[ 9 ][ 11 ]: 1.0
x[ 10 ][ 8 ]: 1.0
x[ 11 ][ 7 ]: 1.0

Constraints
constraints : Size=24
    Key : Lower : Body : Upper
      1 :   1.0 :  1.0 :   1.0
      2 :   1.0 :  1.0 :   1.0
      3 :   1.0 :  1.0 :   1.0
      4 :   1.0 :  1.0 :   1.0
      5 :   1.0 :  1.0 :   1.0
      6 :   1.0 :  1.0 :   1.0
      7 :   1.0 :  1.0 :   1.0
      8 :   1.0 :  1.0 :   1.0
      9 :   1.0 :  1.0 :   1.0
     10 :   1.0 :  1.0 :   1.0
     11 :   1.0 :  1.0 :   1.0
     12 :   1.0 :  1.0 :   1.0
     13 :   1.0 :  1.0 :   1.0
     14 :   1.0 :  1.0 :   1.0
     15 :   1.0 :  1.0 :   1.0
     16 :   1.0 :  1.0 :   1.0
     17 :   1.0 :  1.0 :   1.0
     18 :   1.0 :  1.0 :   1.0
     19 :   1.0 :  1.0 :   1.0
     20 :   1.0 :  1.0 :   1.0
     21 :   1.0 :  1.0 :   1.0
     22 :   1.0 :  1.0 :   1.0
     

#Ans 9:

The Cost is 203.

$x_{1,7}=1$

$x_{2,2}=1$

$x_{3,5}=1$

$x_{4,1}=1$

$x_{5,11}=1$

$x_{6,3}=1$

$x_{7,6}=1$

$x_{8,4}=1$

$x_{9,10}=1$

$x_{10,12}=1$

$x_{11,9}=1$

$x_{12,8}=1$

#Ans 10:

The optimal costs for the both the problems are same. Yes, the values of the variables is still integer valued, it happens because the optimal solution occurs at the extreme points, i.e. at the corners.

In [25]:
coef_reals = np.loadtxt('/content/lab5_ex2_fractional.txt',delimiter = ',')

In [26]:
model_reals.objective_decimals = Objective(expr=sum(coef_reals[i][j]*model_reals.y[i,j] for i in row_indices for j in col_indices),sense = minimize)

In [27]:
model_reals.objective.deactivate()

In [28]:
result_reals = opt_cbc.solve(model_reals)
print('Solver status:', result_reals.solver.status)
print('Solver termination condition:',result_reals.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [29]:
print('Objective = ', model_reals.objective())

print('\nDecision Variables')
for i in row_indices:
  for j in col_indices:
    if model.x[i,j].value == 1:
      print('x[',i,'][',j,']:',model_reals.y[i,j].value)

print('\nConstraints')
model_reals.constraints.display()

Objective =  203.0

Decision Variables
x[ 0 ][ 6 ]: 1.0
x[ 1 ][ 1 ]: 1.0
x[ 2 ][ 4 ]: 1.0
x[ 3 ][ 0 ]: 1.0
x[ 4 ][ 10 ]: 1.0
x[ 5 ][ 2 ]: 1.0
x[ 6 ][ 5 ]: 1.0
x[ 7 ][ 3 ]: 1.0
x[ 8 ][ 9 ]: 1.0
x[ 9 ][ 11 ]: 1.0
x[ 10 ][ 8 ]: 1.0
x[ 11 ][ 7 ]: 1.0

Constraints
constraints : Size=24
    Key : Lower : Body : Upper
      1 :   1.0 :  1.0 :   1.0
      2 :   1.0 :  1.0 :   1.0
      3 :   1.0 :  1.0 :   1.0
      4 :   1.0 :  1.0 :   1.0
      5 :   1.0 :  1.0 :   1.0
      6 :   1.0 :  1.0 :   1.0
      7 :   1.0 :  1.0 :   1.0
      8 :   1.0 :  1.0 :   1.0
      9 :   1.0 :  1.0 :   1.0
     10 :   1.0 :  1.0 :   1.0
     11 :   1.0 :  1.0 :   1.0
     12 :   1.0 :  1.0 :   1.0
     13 :   1.0 :  1.0 :   1.0
     14 :   1.0 :  1.0 :   1.0
     15 :   1.0 :  1.0 :   1.0
     16 :   1.0 :  1.0 :   1.0
     17 :   1.0 :  1.0 :   1.0
     18 :   1.0 :  1.0 :   1.0
     19 :   1.0 :  1.0 :   1.0
     20 :   1.0 :  1.0 :   1.0
     21 :   1.0 :  1.0 :   1.0
     22 :   1.0 :  1.0 :   1.0
     

In [30]:
model.constraints.add(model.x[1,3] == 0)
model.constraints.add(model.x[5,10] == 0)

In [31]:
model.pprint()

4 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   26 : {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain              : Size : Members
        None :     2 : x_index_0*x_index_1 :  144 : {(0, 0), (0, 1), (0, 2), (0, 3), (0, 4), (0, 5), (0, 6), (0, 7), (0, 8), (0, 9), (0, 10), (0, 11), (1, 0), (1, 1), (1, 2), (1, 3), (1, 4), (1, 5), (1, 6), (1, 7), (1, 8), (1, 9), (1, 10), (1, 11), (2, 0), (2, 1), (2, 2), (2, 3), (2, 4), (2, 5), (2, 6), (2, 7), (2, 8), (2, 9), (2, 10), (2, 11), (3, 0), (3, 1), (3, 2), (3, 3), (3, 4), (3, 5), (3, 6), (3, 7), (3, 8), (3, 9), (3, 10), (3, 11), (4, 0), (4, 1), (4, 2), (4, 3), (4, 4), (4, 5), (4, 6), (4, 7), (4, 8), (4, 9), (4, 10), (4, 11), (5, 0), (5, 1), (5, 2), (5, 3), (5, 4), (5, 5), (5, 6), (5, 7), (5, 8), (5, 9), (5, 10

In [32]:
result = opt_cbc.solve(model)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [33]:
print('Objective = ', model.objective())

print('\nDecision Variables')
for i in row_indices:
  for j in col_indices:
    if model.x[i,j].value == 1:
      print('x[',i,'][',j,']:',model.x[i,j].value)

print('\nConstraints')
model.constraints.display()

Objective =  203.0

Decision Variables
x[ 0 ][ 6 ]: 1.0
x[ 1 ][ 1 ]: 1.0
x[ 2 ][ 4 ]: 1.0
x[ 3 ][ 0 ]: 1.0
x[ 4 ][ 10 ]: 1.0
x[ 5 ][ 2 ]: 1.0
x[ 6 ][ 5 ]: 1.0
x[ 7 ][ 3 ]: 1.0
x[ 8 ][ 9 ]: 1.0
x[ 9 ][ 11 ]: 1.0
x[ 10 ][ 8 ]: 1.0
x[ 11 ][ 7 ]: 1.0

Constraints
constraints : Size=26
    Key : Lower : Body : Upper
      1 :   1.0 :  1.0 :   1.0
      2 :   1.0 :  1.0 :   1.0
      3 :   1.0 :  1.0 :   1.0
      4 :   1.0 :  1.0 :   1.0
      5 :   1.0 :  1.0 :   1.0
      6 :   1.0 :  1.0 :   1.0
      7 :   1.0 :  1.0 :   1.0
      8 :   1.0 :  1.0 :   1.0
      9 :   1.0 :  1.0 :   1.0
     10 :   1.0 :  1.0 :   1.0
     11 :   1.0 :  1.0 :   1.0
     12 :   1.0 :  1.0 :   1.0
     13 :   1.0 :  1.0 :   1.0
     14 :   1.0 :  1.0 :   1.0
     15 :   1.0 :  1.0 :   1.0
     16 :   1.0 :  1.0 :   1.0
     17 :   1.0 :  1.0 :   1.0
     18 :   1.0 :  1.0 :   1.0
     19 :   1.0 :  1.0 :   1.0
     20 :   1.0 :  1.0 :   1.0
     21 :   1.0 :  1.0 :   1.0
     22 :   1.0 :  1.0 :   1.0
     

#Ans 12:

We change our pyomo model to introduce the given changes. We add two more constraints, defining $x_{2,4} \text{ and } x_{6,11}$ equal to 0.

The cost is : 203.

$x_{1,7}=1$

$x_{2,2}=1$

$x_{3,5}=1$

$x_{4,1}=1$

$x_{5,11}=1$

$x_{6,3}=1$

$x_{7,6}=1$

$x_{8,4}=1$

$x_{9,10}=1$

$x_{10,12}=1$

$x_{11,9}=1$

$x_{12,8}=1$

We do not see any changes in the solution as even before when facility 2 and 6 could be located at 4 and 11 respectively, they were not.